### ISSUE TITLE
Synchronized Output

### ISSUE THREAD

**Issue by christianparpart** [xtermjs/xterm.js#3375]
Hey guys,

I was searching the issues list (and my email inbox) as I remember I was once in talk with @jerch at lesat about this subject (I think), but I couldn't find it, so I want to ask here officially.

Is it of interest for you to implement the VT mode `?2026` (synchronized output)?
I am currently compiling a list of supporting software (TEs, toolkits, apps) and would like to add you there, too, if that is of interest for you.

What do you think?

**Comment by jerch** [comment]
@christianparpart Yes I am interested in that feature, I think that is a very valuable addition to the terminal interface, esp. for fullpage apps. But I cannot tell if and when we can get to it. The problem I foresee here - we have 3 renderers with very different internal semantics, that all would have to support it (maybe we could de-bounce it at higher level for all, no clue yet).

So plz dont count us in yet, until we had some discussion/thinking about it.

**Comment by Tyriar** [comment]
I can't find docs on this feature but it sounds like it would be really nice if it just prevents rendering until a block is parsed. @jerch if that's how it works we could just delay rendering until after the synchronized block is hit (by skipping the render and re-queueing to next frame), this should be renderer agnostic in RenderService.

**Comment by christianparpart** [comment]
I once summaries that for me, and yesterday (as said) tried to compile a list of supporting software. I've now moved this little document into a dedicated public page: https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036

This does contain the spec as well as the adoption state.

**Comment by jerch** [comment]
@christianparpart Thx for writing down the mechanics of that mode. So DECSET 2026 enters that mode and basically should tell the TE to wait with the screen output until DECRST 2026 or the timeout happens? Or is there some 3rd sequence I have overlooked to mark the inner "video frame" end-start border?

About possible implementation details:
Re-queueing a screen update up to some upper timeout should be easy doable for all renderers in one place (@Tyriar guess we would have to abstract `requestAnimationFrame` to some own impl with a max threshold refer to the real `requestAnimationFrame`). That would be a best-effort solution in terms of multiples of browser FPS, but not a strict sync mode.

But the sync feature raises a basic question in my head - If the sync frame sequence happens within the timeout, should a TE try to render things fully in sync to the band position of the marker? Thus exactly at the time, when that marker is seen by the parser? (No print data missing, no follow-up bytes are shown).

Such a "true sync" mode is quite problematic for xterm.js, due to the inner mechanics: In theory the parser and the renderer run independently in xterm.js (beside the fact they are blocking each other on the mainthread currently, but even that might change in the future with webworkers), the renderer is only snapshotting the buffer from time to time. With a sync marker in the band data we would have to align the buffer state and the renderer tightly, thus either stall the parser until rendering happened (possible with my recent async parser patch) or do some buffer freeze copy to be shown next. Not sure yet if anything that tightly coupled is feasible.

**Comment by christianparpart** [comment]
> Ideally the renderer only depends on that buffer afterwards

That is exactly how i have implemented it. Except that i call the scraping buffer a [render buffer](https://github.com/christianparpart/contour/blob/master/src/terminal/RenderBuffer.h#L26-L51) and that it is actually a double buffer (front/back)  to reduce locking times. When scraping into the render buffer you render into the back buffer and the rendering engine only accesses the front buffer. Thr front/back buffers are swapped based on the screens refresh rate, whether there was any update at all, and now also if updating the back buffer and swapping is permitted (2026 disabled). (See [here](https://github.com/christianparpart/contour/blob/master/src/terminal/Terminal.cpp#L633) and [there](https://github.com/christianparpart/contour/blob/master/src/terminal/Terminal.cpp#L199)).

So yeah, that should decouple the renderer somewhat completely from there terminal as it only needs access to the (scraped) render front buffer.

Another optimization step that proved to be useful: more often than not the visible screen area contains empty grid cells (no text no SGR), these are skipped in my render buffer. So the render buffer only contains cells that do contain something.

I hope that's useful :)

**Comment by jerch** [comment]
@christianparpart Thx for the technical insights. I still stumble over the "mode" aspects of this - if I get it right, there is only an "on" and an "off" sequence, where "off" also means "paint now". Is that enough on TE side, or would it be easier to have a 3rd sequence with the meaning of "paint now", that only would be respected during if 2026 is high? Thus making it a real operation mode on the TE, not just an on/off flag.
A real operation mode might come handy, if a TE has to do more preparation steps, that should stay active between the paint calls. So what is your experience here? Is the on/off flag sufficient?

**Comment by Tyriar** [comment]
FWIW the webgl renderer already uses a double buffer:

https://github.com/xtermjs/xterm.js/blob/d046e2a77f92de89403115e810eeb323a13b17b7/addons/xterm-addon-webgl/src/GlyphRenderer.ts#L20-L26

I think we could always trigger a render immediately when the sync start is hit without that many changes.

**Comment by jerch** [comment]
> FWIW the webgl renderer already uses a double buffer:

Yeah I already suspected that. We def. should try to avoid triple+ buffering where possible :smile_cat:

**Comment by christianparpart** [comment]
I am not sure that's the same thing. Sorry for not knowing your architecture well enough. in OpenGL at least the double buffer is rendered pixels and you swap between them (always rendering to the back), whereas how I implemented Synchronized updates is that the renderer - completely decoupled from the internals of the terminal - only has access to a render buffer that contains some pre-fetch cells from the current view to the terminal's screen (scroll offset applied, selection applied - not sure how applicable that is to you though). so in normal cases, that render buffer is just refreshed on each paintGL-like call. Now, with 2026 mode enabled, fetching the render-buffer is skipped and therefore the paintGL-like call renders the exact same scene as last time, effectively implementing synchronized updates / avoiding tearing.

p.s.: I checked how others are implementing it, and at least Alacritty (though it's using a different VT sequence) seems to implement it via the same strategy. (at least they're also having the same layers).

**Comment by mdc-git** [comment]
https://github.com/mdc-git/SynchronizedOutputAddon

**Comment by Tyriar** [comment]
> https://github.com/mdc-git/SynchronizedOutputAddon

@mdc-git I cannot recommend anyone use that as it monkey patches core methods https://github.com/mdc-git/SynchronizedOutputAddon/blob/5ed222e4cee40b786bfe547e06ad6f8ae9519314/SynchronizedOutputAddon.js#L70-L74

It also breaks behavior like scroll locking with https://github.com/mdc-git/SynchronizedOutputAddon/blob/5ed222e4cee40b786bfe547e06ad6f8ae9519314/SynchronizedOutputAddon.js#L76C1-L83C6

---

Am I missing something or is this basically just about pausing the renderer when that sequence is encountered and then resuming it when its counterpart is hit? If so that's probably like a 20 line change in xterm.js

**Comment by mdc-git** [comment]
@Tyriar 

I addressed your concern. Thanks for the feedback. The monkey patching was more for convenience, but I understand your point (and `ingest` was already there anyway). I also removed the bottom scroll because I can use `scrollOnUserInput` again. You're right, the implementation of the spec alone shouldn't require much code. In this case, it's more of a byproduct. However, as outlined in the README, there are a number of other visual annoyances, even after implementing the spec, as seen in other terminals with Claude Code specifically. The add-on was specifically designed to address these issues.  

EDIT:

~~I removed most of the bloat as well.~~ Didn't have the expected outcome.

**Comment by chrislloyd** [comment]
@Tyriar I've put up a PR for this at #5453 if you want to take a look - would love to get this over the line!

**Comment by mdc-git** [comment]
[Finally](https://github.com/anthropics/claude-code/issues/769#issuecomment-3667315590) this got some traction. Thank you.
